File: app/src/main/java/com/vladislav/workoutassistant/exercises/adapters/ExerciseAdapter.java
package com.vladislav.workoutassistant.exercises.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.vladislav.workoutassistant.R;
import com.vladislav.workoutassistant.core.DiffUtilCallback;
import com.vladislav.workoutassistant.core.callbacks.ItemClickCallback;
import com.vladislav.workoutassistant.data.db.entity.Exercise;
import com.vladislav.workoutassistant.databinding.ItemExerciseBinding;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

public class ExerciseAdapter extends RecyclerView.Adapter<ExerciseAdapter.ExerciseViewHolder>{

    private List<Exercise> mExercises;
    private ItemClickCallback mCallback;
    private boolean mIsMultipleSelectionMode;

    public ExerciseAdapter(ItemClickCallback callback, boolean isMultipleSelectionMode) {
        mCallback = callback;
        mIsMultipleSelectionMode = isMultipleSelectionMode;
    }

    public void updateList(List<Exercise> exercises) {
        if (mExercises == null) {
            mExercises = exercises;
            notifyItemRangeInserted(0, exercises.size());
        } else {
            DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtilCallback(mExercises, exercises));
            mExercises = exercises;
            diffResult.dispatchUpdatesTo(this);
        }
    }

    @Override
    @NonNull
    public ExerciseAdapter.ExerciseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemExerciseBinding binding = DataBindingUtil.inflate(inflater, R.layout.item_exercise, parent, false);
        return new ExerciseAdapter.ExerciseViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ExerciseAdapter.ExerciseViewHolder holder, int position) {
        holder.bind(mExercises.get(position));
    }

    @Override
    public int getItemCount() {
        return mExercises == null ? 0 : mExercises.size();
    }

    class ExerciseViewHolder extends RecyclerView.ViewHolder {
        private ItemExerciseBinding mBinding;
        private Exercise mExercise;

        ExerciseViewHolder(ItemExerciseBinding binding) {
            super(binding.getRoot());

            mBinding = binding;
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mIsMultipleSelectionMode) {
                        mBinding.setIsSelected(true);
                    }
                    mCallback.onClick(mExercise.getId(), mExercise.getName());
                }
            });
//            mBinding.setCallback(mCallback);
        }

        void bind(Exercise exercise) {
            mExercise = exercise;
            mBinding.setId(exercise.getId());
            mBinding.setName(exercise.getName());
        }
    }
}

File: feature-exercise-book-impl/src/main/java/com/vladislavmyasnikov/feature_exercise_book_impl/presentation/adapters/ExerciseAdapter.kt
package com.vladislavmyasnikov.feature_exercise_book_impl.presentation.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.chip.Chip
import com.vladislavmyasnikov.core_components.interfaces.OnItemClickCallback
import com.vladislavmyasnikov.core_utils.utils.utils.DiffUtilCallback
import com.vladislavmyasnikov.feature_exercise_book_impl.R
import com.vladislavmyasnikov.feature_exercise_book_impl.domain.ShortExerciseInfo
import kotlinx.android.synthetic.main.item_exercise.view.*
import javax.inject.Inject

class ExerciseAdapter @Inject constructor(private val context: Context) : RecyclerView.Adapter<ExerciseAdapter.ViewHolder>() {

    var callback: OnItemClickCallback? = null

    private var items: List<ShortExerciseInfo> = emptyList()
    private val muscleGroups: Array<String> = context.resources.getStringArray(R.array.muscle_groups)

    fun updateList(_items: List<ShortExerciseInfo>) {
        val diffResult = DiffUtil.calculateDiff(DiffUtilCallback(items, _items))
        items = _items
        diffResult.dispatchUpdatesTo(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.item_exercise, parent, false)
        val holder = ViewHolder(view, muscleGroups, context)

        holder.itemView.setOnClickListener {
            val item = items[holder.adapterPosition]
            callback?.onClick(item.id, item.title)
        }
        return holder
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(items[position])
    }

    override fun getItemCount(): Int = items.size

    class ViewHolder(view: View, private val muscleGroups: Array<String>, context: Context) : RecyclerView.ViewHolder(view) {

        init {
            for (i in 0 until muscleGroups.size) {
                val tag = Chip(context).apply { visibility = View.GONE }
                itemView.muscle_groups_tags.addView(tag)
            }
        }

        fun bind(item: ShortExerciseInfo) {
            itemView.title_field.text = item.title
            for ((position, id) in item.muscleGroupsIDs.withIndex()) {
                val tag = itemView.muscle_groups_tags.getChildAt(position) as Chip
                tag.text = muscleGroups[id]
                tag.visibility = View.VISIBLE
            }
            for (i in item.muscleGroupsIDs.size until muscleGroups.size) {
                val tag = itemView.muscle_groups_tags.getChildAt(i) as Chip
                tag.visibility = View.GONE
            }
        }
    }
}